File: Go/qg.go
// Comment line
/*
Comment block
*/

package main // Used as "main" package

import (
	"fmt"
) // Import "fmt" package

func main() {
	// Declare variables
	var var1 string = "abc" // Type is explicitly declared
	var var2 = "def"        // Type is inferred
	var3 := "ghi"           // Type is inferred, this declare method only can be used inside function

	var var4 string // No type is declared
	var4 = "jkl"    //Now the variable is initialized with type

	// Decalre multiple variables
	var var5, var6 int = 5, -6 // Same type
	var var7, var8 = "7", 8    // Different types
	var9, var10 := "9", 10.0
	var ( // Declare in a block
		var11 int
		var12 = "12"
	)

	fmt.Println("Variables declared:\n", var5, var6, var7, var8, var9, var10, var11, var12) // All variables must be used for output

	fmt.Print("\n")

	// Constants
	const CONST1 string = "constant" // Should name it in FULL UPPERCASE to differentiate
	const CONST2 = 0                 // Should name it in FULL UPPERCASE to differentiate
	const (                          // Declare in block
		CONST3 int = 3
		CONST4     = -4
	)

	// Variable types
	var (
		intVar    int     = 20
		floatVar  float32 = 12.34
		stringVar string  = "string"
		boolVar   bool    = true
	)

	// Console print, using "fmt" package
	fmt.Print(var1, var2, var3)                           // No whitespace and newline added
	fmt.Println(var1, var2, var3)                         // Whitespace added between arguments and newline added at the end
	fmt.Printf("var4\nvalue: %v\ntype: %T\n", var4, var4) // Print with format, no newline

	fmt.Print("\n")

	// Print format
	// 1. General formatting verbs
	fmt.Printf("%v\n", var7) // Print value
	fmt.Printf("%v\n", var8)
	fmt.Printf("%#v\n", var7) // Print value in Go-syntax format
	fmt.Printf("%#v\n", var8)
	fmt.Printf("%v%%\n", var7) // Print with % sign
	fmt.Printf("%v%%\n", var8)
	fmt.Printf("%T\n", var7) // Print type of the value
	fmt.Printf("%T\n", var8)

	// 2. Integer formattting verbs
	fmt.Printf("%b\n", intVar)   // as binary, base2
	fmt.Printf("%d\n", intVar)   // as decimal, base10
	fmt.Printf("%+d\n", intVar)  // as decimal with sign(+/-)
	fmt.Printf("%o\n", intVar)   // as octal, base8
	fmt.Printf("%O\n", intVar)   // as octal with 0o
	fmt.Printf("%x\n", intVar)   // as lowercase haXadecimal, base 16
	fmt.Printf("%X\n", intVar)   // as uppercase haXadecimal
	fmt.Printf("%#x\n", intVar)  // haxadecimal with 0x
	fmt.Printf("%4d\n", intVar)  // pad with spaces, right justified
	fmt.Printf("%-4d\n", intVar) // pad with spaces, left justified
	fmt.Printf("%04d\n", intVar) // pad with zeroes

	fmt.Print("\n")

	// 3. Float formatting verbs
	fmt.Printf("%e\n", floatVar)    // as scientific notation, with exponent
	fmt.Printf("%f\n", floatVar)    // as decimal point, without exponent
	fmt.Printf("%.2f\n", floatVar)  // with precision
	fmt.Printf("%6.2f\n", floatVar) // pad with spaces, right justified
	fmt.Printf("%g\n", floatVar)    // exponent as needed

	fmt.Print("\n")

	// 4. String formatting verbs
	fmt.Printf("%s\n", stringVar)   // as plain string
	fmt.Printf("%q\n", stringVar)   // as double-quoted string
	fmt.Printf("%8s\n", stringVar)  // pad with spaces, right justified
	fmt.Printf("%-8s\n", stringVar) // pad with spaces, left justified
	fmt.Printf("%x\n", stringVar)   // as hex dump
	fmt.Printf("% x\n", stringVar)  // as hex dump, spaces between hex

	fmt.Print("\n")

	// 5. Boolean formatting verbs
	fmt.Printf("%t\n", boolVar) // Print value

	fmt.Print("\n")

	// Array
	var array1 = [3]string{"a", "b", "c"}   // Explicitly declare length
	var array2 = [...]string{"d", "e", "f"} // Inferredly declare length

	fmt.Println(array1[0]) // Access element
	array2[1] = "E"        // Change element

	fmt.Print("\n")

	var array3 = [5]string{"g", "h", "i"} // Fill empty elements with 0 or ""
	fmt.Println(array3, "\n")             // Print whole array, notice default value

	var array4 = [3]int{1: 8, 2: 4} // Initialize specific elements
	fmt.Println(array4, "\n")

	fmt.Printf("Lengths\narray1:%v\narray2:%v\narray3:%v\narray4:%v\n\n", len(array1), len(array2), len(array3), len(array4))

	// Slices, similar to array but dynamic size
	var slice1 = []int{}        // Declare empty slice
	var slice2 = []int{1, 2, 3} // Declare slice with initialization

	fmt.Println("slice1: ", slice1)
	fmt.Println("slice1 length: ", len(slice1))         // Show slice length
	fmt.Println("slice1 capacity: ", cap(slice1), "\n") // Show slice capacity

	fmt.Println("slice2: ", slice2)
	fmt.Println("slice2 length: ", len(slice2))
	fmt.Println("slice2 capacity: ", cap(slice2), "\n")

	var slice3 = array3[1:2] // Create slice from array, [fromIndex: beforeIndex]
	fmt.Println("slice3: ", slice3)
	fmt.Println("slice3 length: ", len(slice3))
	fmt.Println("slice3 capacity: ", cap(slice3), "\n")

	var slice4 = make([]string, 3, 6) // make([]type, length, capacity), capacity = length if not defined
	fmt.Println("slice4: ", slice4)
	fmt.Println("slice4 length: ", len(slice4))
	fmt.Println("slice4 capacity: ", cap(slice4), "\n")

	slice4[2] = "j"                     // change slice element
	slice4 = append(slice4, "k", "l")   // update slice by appending new elements
	slice5 := append(slice3, slice4...) // "..." is always needed
	fmt.Println("slice4: ", slice4)
	fmt.Println("slice5: ", slice5, "\n")

	slice5 = slice5[3:4] // reduce size by slicing
	fmt.Println("slice5: ", slice5)
	fmt.Println("slice5 length: ", len(slice5))
	fmt.Println("slice5 capacity: ", cap(slice5), "\n")

	slice5 = append(slice5, "m", "n") // adding size by appending
	fmt.Println("slice5: ", slice5)
	fmt.Println("slice5 length: ", len(slice5))
	fmt.Println("slice5 capacity: ", cap(slice5), "\n")

	slice6 := []string{"o", "p", "q"}
	copy(slice6, slice5[1:3]) // copy(dest, src), copy elements from "src" to "dest" until no elements
	// useful for using elements from large array/slice
	fmt.Println("slice6: ", slice6)
	fmt.Println("slice6 length: ", len(slice6))
	fmt.Println("slice6 capacity: ", cap(slice6), "\n")

	// Operators
	var num1 = 2
	num2 := 4
	var (
		num3 = 6
		num4 = 8
	)
	// 1. Arithmatic
	fmt.Println(num1 + num2)
	fmt.Println(num1 - num2)
	fmt.Println(num1 * num2)
	fmt.Println(num1 / num2)
	fmt.Println(num1 % num2)
	num1++
	fmt.Println(num1)
	num1--
	fmt.Println(num1, "\n")
	// 2. Comparison
	fmt.Println(num3 == num3)
	fmt.Println(num3 != num4)
	fmt.Println(num4 > num3)
	fmt.Println(num3 < num4)
	fmt.Println(num3 >= num3)
	fmt.Println(num3 <= num3, "\n")
	// 3. Logical
	fmt.Println(num3 == num3 && num3 == num3)
	fmt.Println(num3 == num4 || num3 == num3)
	fmt.Println(!(num3 == num3), "\n")
	// 4. Bitwise
	fmt.Println(num1 & num1)
	fmt.Println(num1 | num1)
	fmt.Println(num1 ^ num1)
	fmt.Println(num1 << 2)
	fmt.Println(num1>>2, "\n")
	// 5. Assignment
	num5 := 10
	num6 := 12
	num5 += num6
	fmt.Println(num5)
	num5 -= num6
	fmt.Println(num5)
	num5 *= num6
	fmt.Println(num5)
	num5 /= num6
	fmt.Println(num5)
	num5 %= num6
	fmt.Println(num5, "\n")
	num7 := 2
	num8 := 4
	num7 &= num8
	fmt.Println(num7)
	num7 |= num8
	fmt.Println(num7)
	num7 ^= num8
	fmt.Println(num7)
	num7 >>= 2
	fmt.Println(num7)
	num7 <<= 2
	fmt.Println(num7, "\n")

	// Conditional statement
	stringVar2 := "B"
	if stringVar2 == "A" {
		fmt.Println("stringVar2 is A")
	} else if stringVar2 == "B" {
		fmt.Println("stringVar2 is B")
	} else {
		fmt.Println("stringVar2 is something else")
	}

	fmt.Println()

	switch stringVar2 {
	case "A":
		fmt.Println("stringVar2 is A")
	case "B":
		fmt.Println("stringVar2 is B")
	case "C":
		fmt.Println("stringVar2 is C")
	case "D", "E", "F":
		fmt.Println("stringVar2 is D, E, or F")
	default:
		fmt.Println("stringVar2 is something else")
	}

	fmt.Println()

	// Loops
	var i = 0
	for i < 10 { // golang does not have while loop, this is the substitude
		fmt.Println("Golang doesnt have while loop")
		i++
	}

	fmt.Println()

	for i := 0; i <= 10; i++ { // for loop
		if i == 4 {
			fmt.Println()
			continue
		} else if i == 10 {
			break
		} else {
			fmt.Printf("loop no.%v\n", i)
		}
	}

	fmt.Println()

	array5 := []int{2, 4, 6, 8, 10}
	for idx, val := range array5 { // iterate elements [for index, value := range array/slice/map]
		fmt.Printf("%v |\t%v\n", idx, val)
	}

	fmt.Println()

	for _, val := range array5 { // "_" will be omited
		fmt.Println(val)
	}

	// Call functions [ref1]
	func1()
	func2("text", 0) // put arguments
	var (            // store return values
		resultFunc3                = func3(1, 2)
		resultFunc4                = func4()
		resultFunc5a, resultFunc5b = func5()
	)
	fmt.Println()
	fmt.Printf("Results\nresultFunc3: %v\nresultFunc4: %v\nresultFunc5: %v, %v\n\n", resultFunc3, resultFunc4, resultFunc5a, resultFunc5b)

	// Structures, object-like
	type structType1 struct { // Declare a structure
		member1 string
		member2 int
		member3 bool
	}
	var structVar1 structType1 // use default value if no initialization
	fmt.Println(structVar1)
	structVar1.member1 = "Member1" // access struct members
	fmt.Println(structVar1.member1, "\n")

	// Maps, key-value pairs
	var mapVar1 = map[string]int{"First": 20, "Second": 40} // init map with string:int pair
	var mapVar2 = map[int]string{0: "Car", 1: "Van"}        // init map with int:string pair
	var mapVar3 = make(map[int]bool)                        // make empty int:bool map variable without init
	var mapVar4 map[int]int                                 // create empty int:int, not encourage as can cause panic/error
	fmt.Println("Map variables:")
	fmt.Println(mapVar1)
	fmt.Println(mapVar2)
	fmt.Println(mapVar3)
	fmt.Println(mapVar4, "\n")

	fmt.Print("mapVar1[\"First\"]: ")
	fmt.Println(mapVar1["First"]) // access map elements
	fmt.Print("mapVar2[0]: ")
	fmt.Println(mapVar2[0], "\n")

	fmt.Print("mapVar1[\"A\"]: ")
	fmt.Println(mapVar1["A"]) // default value will given if not found
	fmt.Print("mapVar2[100]: ")
	fmt.Println(mapVar2[100], "\n")

	mapVar1["Third"] = 60 // change map element, add if not found
	fmt.Print("mapVar1[\"Third\"]: ")
	fmt.Println(mapVar1["Third"], "\n")

	delete(mapVar1, "Third") // remove map element
	fmt.Print("mapVar1[\"Third\"]: ")
	fmt.Println(mapVar1["Third"], "\n")

	mapVar1Third, mapVar1ThirdExist := mapVar1["Third"] // first var holds value, second var hold bool for existance
	fmt.Println("mapVar1[\"Third\"]:", mapVar1Third, "| If exist:", mapVar1ThirdExist, "\n")

	_, mapVar1FourthExist := mapVar1["Fourth"] // "_" will be omited
	fmt.Println("If mapVar1[\"Fourth\"] exist:", mapVar1FourthExist, "\n")

	var mapVar5 = mapVar1 // maps are references
	fmt.Println("Before")
	fmt.Println("mapVar1:", mapVar1)
	fmt.Println("mapVar5:", mapVar5)
	mapVar1["Third"] = 100 // they share same hash table, they will change together
	fmt.Println("After")
	fmt.Println("mapVar1:", mapVar1)
	fmt.Println("mapVar5:", mapVar5, "\n")

	fmt.Println("Iterate mapVar1: ", mapVar1) // map can be iterated over, but not always in order
	for key, val := range mapVar1 {
		fmt.Printf("%v: %v\n", key, val)
	}

	fmt.Println()

	var mapVar1Order []string
	mapVar1Order = append(mapVar1Order, "Third", "Second", "First") // define an array of keys for order
	fmt.Println("Iterate mapVar1: ", mapVar1)
	for _, mapVar1Idx := range mapVar1Order {
		fmt.Printf("%v: %v\n", mapVar1Idx, mapVar1[mapVar1Idx])
	}

}

// Create functions [ref1]
func func1() {
	fmt.Println("func1")
}
func func2(input1 string, input2 int) { // with multiple parameters
	fmt.Println("func2:", input1, input2)
}
func func3(input1 int, input2 int) int { // with single return value
	return input1 + input2
}
func func4() (result string) { // with named return value
	result = "function 4 have a named return value"
	return
}
func func5() (string, string) { // with multiple return values, can be named as well
	return "multiple", "results"
}

// https://www.w3schools.com/go/go_maps.php

File: PHP/qg.php
<?php declare(strict_types=1);    // set PHP to become type-strict, must be first PHP statement
# PHP single line comment
// PHP single line comment
/* PHP multiline
comment*/

// Echo to print output into console
echo "Hello World\n";
echo phpversion(), "\n";
echo "\n";

// Declaring variables, PHP is loosely typed, no need type declaration
$var1 = "String";   // string
$var2 = 10;         // int
$var3 = 12.34;      // float
$var4 = true;       // bool
$var5 = null;       // NULL

echo "\$var1: ", $var1, "\n";
echo "\$var2: ", $var2, "\n";
echo "\$var3: ", $var3, "\n";
echo "\$var4: ", $var4, "\n";
echo "\$var5: ", $var5, "\n";
echo "\n";

// Return the type of variables
var_dump($var1);
var_dump($var2);
var_dump($var3);
var_dump($var4);
var_dump($var5);
echo "\n";

// Assign same value on multiple variables
$var6 = $var7 = "Same";
echo "\$var6: ", $var6, "\n";
echo "> ";
var_dump($var6);
echo "\$var7: ", $var7, "\n";
echo "> ";
var_dump($var7);
echo "\n";

// Console echo
echo "Echo without parenthesis\n";
echo("Echo with parenthesis\n");
echo "\n";

// Console print vs echo
$printVar = print "Print returns status code '1'\n";
echo "\$printVar: ", $printVar, "\n";
echo "Echo doesnt return\n";
echo "\n";

// Echo/print often used to output HTML
echo "<h1>This is PHP echo output<h1>";
echo "\n\n";

// Variable in string
$var8 = "simple";
$var9 = "This is $var8";  // Double-quote string will process special characters, can directly insert variable
echo "\$var8: ", $var8, "\n";
echo "\$var9: ", $var9, "\n";
echo "\n";

$var10 = "complex";
$var11 = 'This is '. $var10;  // Single-quote string is literal, need concatenation to insert variable 
echo "\$var10: ", $var10, "\n";
echo "\$var11: ", $var11, "\n";
echo "\n";

// String count
echo 'strlen("123456789") = '. strlen("123456789"), "\n";   // strlen() return string length
echo 'str_word_count("Two words") = '. str_word_count("Two words"), "\n";   // str_word_count() return word count
echo "\n";

// String search
$var12 = "tryFind waldoInthis weirdstring";
echo 'strpos($var12, "waldo") = ', strpos($var12, "waldo"), "\n";   // strpos() find needle (arg2) in haystack (arg1), return 1st char index
echo "\n";

$ifWaldoFound = strpos($var12, "Waldo");
echo 'strpos($var12, "Waldo") = ', $ifWaldoFound, "\n";
echo "\$ifWaldoFound: ", $ifWaldoFound, "\n";
echo "> ";
var_dump($ifWaldoFound);   // return False if not found
echo "\n";

// String manipulation
$oriStr = " Original String !! ";
echo $oriStr, "\n";
echo strtoupper($oriStr), "\n"; // to UPPERCASE
echo strtolower($oriStr), "\n"; // to lowercase
echo str_replace("Original","Replaced",$oriStr), "\n"; // replace substring
echo strrev($oriStr), "\n"; // reverse string
echo trim($oriStr), "\n"; // trim whitespaces
$bitsStr = explode(" ",$oriStr);  // explode string into array
echo "\n";

// String concatenation
$strVar1 = "Happy" . "Birthday";
echo $strVar1, "\n";   // using dot
echo "\n";

// String slicing
echo substr($oriStr, 10, 6), "\n";  // slice (arg1) from index (arg2, not inclusive) to (arg3) characters 
echo substr($oriStr, 10), "\n";  // empty (arg3) means to infinite
echo substr($oriStr, -10, 6), "\n";  // negative (arg2) means slicing from end
echo substr($oriStr, 10, -4), "\n";  // slice to end index (arg3)
echo "\n";

// Escape special characters
echo "\\\" = \"Double quotes\"", "\n";
echo '\\\' = \'Single quotes\'', "\n";
echo "\\\$ = \$ PHP variable", "\n";
echo "\\n = \n new line", "\n";
echo "\\t = \t tab", "\n";
echo "\\ooo = \200 octal value", "\n";
echo "\\xhh = \x11 hex value", "\n";
echo "\n";

// Numbers
$intVar = 12;
$floatVar = 1.234;
$infVar = 1.5e600;  // shown as INF if value is too large
$nanVar = acos(20); // shown as NAN if value is invalid operation
echo "\$intVar: ", $intVar, "\n";
echo "> ";
var_dump($intVar);
echo "\$infVar: ", $infVar, "\n";
echo "> ";
var_dump($infVar);
echo "\$floatVar: ", $floatVar, "\n";
echo "> ";
var_dump($floatVar);
echo "\$nanVar: ", $nanVar, "\n";
echo "> ";
var_dump($nanVar);
echo "\n";

// Check if variable is number/numeric
$numVar1 = 10;
$numVar2 = 20;
$numVar3 = "30";
$numVar4 = "30OO";
var_dump(is_numeric($numVar1));
var_dump(is_numeric($numVar2));
var_dump(is_numeric($numVar3));
var_dump(is_numeric($numVar4));
echo "\n";

// Type casting, cast by putting bracket with a target type in front of the variable
$oriVar = 1;
var_dump((string) $oriVar);
var_dump((int) $oriVar);
var_dump((float) $oriVar);
var_dump((bool) $oriVar);
echo "\n";

// PHP contains mathematic functions
echo pi(), "\n";
echo min(0, 150, 30, 20, -8, -200), "\n";
echo max(0, 150, 30, 20, -8, -200), "\n";
echo abs(-6.7), "\n";
echo sqrt(64), "\n";
echo round(0.60), "\n";
echo round(0.49), "\n";
echo rand(), "\n";
echo rand(10, 100), "\n";    // random between (arg1) and (arg2), both inclusive
echo "\n";

// Declare constants
const CONST_INT1 = 100; // cannot create inside block scope
define("CONST_INT2", 200);  // always global
echo CONST_INT1, "\n";
echo CONST_INT2, "\n";
echo "\n";

// Operators
// 1. arithmatic
echo $numVar1 + $numVar2, "\n";
echo $numVar1 - $numVar2, "\n";
echo $numVar1 * $numVar2, "\n";
echo $numVar1 / $numVar2, "\n";
echo $numVar1 % $numVar2, "\n";
echo $numVar1 ** $numVar2, "\n";
echo "\n";
// 2. increments
echo ++$numVar1, "\n";
echo $numVar1++, "\n";
echo --$numVar1, "\n";
echo $numVar1--, "\n";
echo "\n";
// 3. assignment
$oriVar = $numVar1;
echo $oriVar, "\n";
$oriVar += $numVar1;
echo $oriVar, "\n";
$oriVar -= $numVar1;
echo $oriVar, "\n";
$oriVar *= $numVar1;
echo $oriVar, "\n";
$oriVar /= $numVar1;
echo $oriVar, "\n";
$oriVar %= $numVar1;
echo $oriVar, "\n";
$oriVar .= $numVar1;
echo $oriVar, "\n";
echo "\n";
// 4. comparison
var_dump($numVar1 == $numVar2);
var_dump($numVar1 === $numVar2);
var_dump($numVar1 != $numVar2);
var_dump($numVar1 <> $numVar2);  // same as !=
var_dump($numVar1 > $numVar2);
var_dump($numVar1 >= $numVar2);
var_dump($numVar1 < $numVar2);
var_dump($numVar1 <= $numVar2);
var_dump($numVar1 <=> $numVar2); // spaceship, -1 = less then, 0 = equal, 1 = greater than
echo "\n";
// 5. logical
$true = true;
$false = false;
var_dump($true and $true);
var_dump($true && $true);
var_dump($true or $false);
var_dump($true || $false);
var_dump($true xor $false);
var_dump(!$true);
echo "\n";
// 6. conditional assignment
$result1 = $true ? "true" : "false"; // varName = statement ? value if true : value if false;
echo $result1, "\n";
$result2 = "not NULL" ?? "NULL"; // varName = value ?? if value is null;
echo $result2, "\n";
echo "\n";

// Arrays
$arrVar1 = array(1, 2, 3);  // declare array, this is indexed array
$arrVar2 = ["A", "B", "C",];    // shorthand way to declare
$arrVar3 = ["One"=>1, "Two"=>2, "Three"=>"3",]; // associative array, key-value pair
$arrVar4 = [0=>"1", 1=>"2", 2=>"3",];   // array works like this actualy
$arrVar4 = [true=>"1", "1"=>"2", 2=>"3", "Four"=>null, null=>6, 400];   // all key and value can be different data types, unnamed key will be default to index
var_dump($arrVar1); // proper way to output whole array
var_dump($arrVar2);
var_dump($arrVar3);
var_dump($arrVar4);
echo "\n";
echo '$arrVar1[0] = ', $arrVar1[0], "\n"; // access element in index array
echo '$arrVar4[true] = ', $arrVar4[true], "\n"; // access element in associative array
echo '$arrVar4[1] = ', $arrVar4[1], "\n"; // if the key is numeric, it can be called in integer or string
echo '$arrVar4["2"] = ', $arrVar4["2"], "\n";
echo '$arrVar4["Four"] = ', $arrVar4["Four"], "\n";
echo '$arrVar4[\'Four\'] = ', $arrVar4['Four'], "\n";   // key can be called using single or double quote
echo '$arrVar4[null] = ', $arrVar4[null], "\n";
echo "\n";
$arrVar4[null] = 500;    // update element value
$arrVar4[10] = 5;    // add new if not exist
array_push($arrVar4, "600", "700");    // push new element at the array end, notice its index is behind previous largest
$arrVar4[] = 800;    // shorthand way to push
$arrVar4 += [900, "thousand"=>1000];    // append new array
var_dump($arrVar4);
echo "\n";
array_splice($arrVar1,1,1); // remove (arg3) elements starting index (arg2) in array (arg1), indexed array will reindex
var_dump($arrVar1);
array_splice($arrVar4,1,1); // number-key treated as index and will reindex
var_dump($arrVar4);
unset($arrVar2[1]); // unset will remove element without reindex
var_dump($arrVar2);
$arrVar2 = array_diff($arrVar2, ["C"]); // array_diff() to return the difference between arrays but use it to remove elements
var_dump($arrVar2);
array_pop($arrVar4);    // pop last element
array_shift($arrVar4);  // remove first element
var_dump($arrVar4);
echo "\n";
sort($arrVar1); // sort in ascending order
var_dump($arrVar1);
rsort($arrVar1); // sort in decensing order
var_dump($arrVar1);
asort($arrVar4); // sort in ascending order, based on value
var_dump($arrVar4);
arsort($arrVar4); // sort in ascending order, based on value
var_dump($arrVar4);
ksort($arrVar4); // sort in ascending order, based on value
var_dump($arrVar4);
krsort($arrVar4); // sort in ascending order, based on value
var_dump($arrVar4);
echo "\n";

// Conditional Statements
// 1. if-elseif-else
if ($true) {
    echo "This is true \n";
}
elseif ($false) {
    echo "This is false \n";
}
else {
    echo "This is error \n";
}
echo "\n";
// 2. shorthand if
if ($true) echo "This is true \n";
echo "\n";
// 3. ternary operators
$result1 = $true ? "true" : "false";
echo $result1, "\n";
echo "\n";
// 4. switch
switch ($numVar1 <=> $numVar2) {
    case -1:
        echo "\$numVar1 < \$numVar2 \n";
        break;
    case 0:
    case 1:
        echo "\$numVar1 >= \$numVar2 \n";
        break;
    default:
        echo "This is error \n";
        break;
}
echo "\n";

// Loops
// 1. while
$i = 0;
while ($i < 5) {
    echo "\$i = $i\n";
    $i++;
}
echo "\n";
// 2. do-while
$i = 0;
do {
    echo "\$i = $i\n";
    $i++;
} while ($i < 5);
echo "\n";
// 3. for
for ($i = 0; $i < 5; $i++) {
    if ($i == 2) {
        continue;   // skips the current iteration
    }
    echo "\$i = $i\n";
    if ($i == 4) {
        break;  // break out the loop
    }
}
echo "\n";
// 4. foreach, loop through arrays
foreach ($arrVar1 as $x) {  // loop through indexed arr
    echo '$x = ', "$x", "\n";
}
echo "\n";
foreach ($arrVar4 as $k => $v) {    // loop through associative arr
    echo "$k => $v", "\n";
}
echo "\n";
foreach ($arrVar1 as &$x) {  // set reference variable to refer the originals, easy to update value in foreach
    $x = 1;
}
var_dump($arrVar1);
echo "\n";

// Functions
function funcVar1() {    // declare a function
    echo "funcVar1() called", "\n";
    echo "\n";
}
funcVar1();  // call a function

function funcVar2($param1, $param2) {   // with multiple arguments
    echo "funcVar2() called with param [$param1, $param2]", "\n";
    echo "\n";
}
funcVar2(1, "2");

function funcVar3($param1 = 50) {   // default argument value
    echo "funcVar3() called with param [$param1]", "\n";
    echo "\n";
}
funcVar3();
funcVar3(1);

function funcVar4() {   // with return value
    echo "funcVar4() returns this", "\n";
    echo "\n";
    return "Hi";
}
echo 'funcVar4() = ', funcVar4(), "\n";
echo "\n";

function funcVar5(&$refer) {   // argument by reference
    $refer *= 100;
}
$numVar1 = 10;
echo '$numVar1 = ', $numVar1, "\n";
funcVar5($numVar1);
echo '$numVar1 = ', $numVar1, "\n";
echo "\n";

function funcVar6($param1, ...$params) {   // with variadic arguments (only as last arg), making it variadic function
    echo '$param1 = ', $param1, "\n";
    foreach ($params as $x => $y) {
        echo "$x => $y", "\n";
    }
    echo "\n";
}
funcVar6(1, 2, "3", true, null);

function funcVar7(int $param1) : int {   // with strict_type on, PHP can explicitly typed
    return $param1 + "2";
}
echo 'funcVar7() = ', funcVar7(1), "\n";
echo "\n";

//https://www.w3schools.com/php/php_magic_constants.asp
?>